Documentation: grpc example

pcj opened the pull request:
Closes #2284

pcj commented:
/assign @aledbf

aledbf commented:
@pcj thank you for this example. Please check the headers of the `.go` files to pass the validations.

About the PR:
1. Please remove the installation steps (and k8s manifest). Just assume the cluster is up and running with the ingress controller installed. We already tried this approach before and is really hard to maintain
2. You are adding too many requirements for the example. For instance I would prefer to build and host the docker image in quay.io with the ingress-nginx image and not require the use of gcloud registry
3. Same comment for the dns requirement. You are adding too many references to gcloud and some user can think this only works in that cloud provider.
4. I like the use of bazel but again most of the user just need grpcurl to test this. Can you just use that?

pcj commented:
thanks @aledbf I'll fix up the go headers.  So that I understand you correctly:

1. Remove the nginx-ingress/* directory completely, replacing it with a paragraph in the README.md to "install nginx-ingress in standard fashion";

2. Remove references to gcloud, replacing them with statements like "have persistent disk allocated", or something like that.

3. Remove dns instructions, replacing it with again a paragraph to the effect "have a DNS pointing to your ingress controller".

4. Remove all the bazel stuff, including the build steps for all the images, and all the clients.  Also remove the fortune-teller application (?), replacing with an paragraph like "get your gRPC application up and running..".

What about the kube-cert-manager? 

Anyway, I guess I should have asked this already, but what is your vision of the files necessary to have a functioning gRPC example?  The one I have provided is pretty complete.  So, what is the best combination of maintainability but still approachable.  An example with multiple magical leaps within it will be pretty frustrating for people wanting to try it out.

Thanks for the prompt review!  I'll try to keep to the same standard but I have to steal time away from other commitments, but I'll do my best.

aledbf commented:
> Remove the nginx-ingress/* directory completely, replacing it with a paragraph in the README.md to "install nginx-ingress in standard fashion";

Yes

> Remove references to gcloud, replacing them with statements like "have persistent disk allocated", or something like that.

> Remove dns instructions, replacing it with again a paragraph to the effect "have a DNS pointing to your ingress controller".

Yes

> Remove all the bazel stuff, including the build steps for all the images, and all the clients. Also remove the fortune-teller application (?), replacing with an paragraph like "get your gRPC application up and running..".

No. Just use grpcurl to show the example working. We can publish your image in quay.io so the user doesn't need to build a custom one and have a reference to your code to see how to build a gRPC app

> What about the kube-cert-manager?
> Anyway, I guess I should have asked this already, but what is your vision of the files necessary to have a functioning gRPC example? The one I have provided is pretty complete. 

Agree. Your example is self-contained which I like but you assume too much about the environment where the cluster is running.

> So, what is the best combination of maintainability but still approachable? An example with multiple magical leaps within it will be pretty frustrating for people wanting to try it out.

Agree. That is why I just said you should use the default user guide and avoid using too many references to a particular cloud provider. Keep in mind this must work OOTB with minikube too.

aledbf commented:
@pcj ping

pcj commented:
Hi @aledbf sorry for delay, have not lost interest here...  traveling now but I think I can get this taken care of by Friday or Saturday.

pcj commented:
@aledbf Apologies for delay.  Example is now just a README.md and a few yaml files.  I moved implementation of the grpc service to a separate repo and pushed that to a container registry.

aledbf commented on docs/examples/grpc/ingress.yaml:
```diff
@@ -0,0 +1,21 @@
+metadata:
+  annotations:
+    kubernetes.io/ingress.class: "nginx"
+    ingress.kubernetes.io/ssl-redirect: "true"
```
nginx prefix missing

pcj replied:
Done

aledbf commented:
@pcj just two comments. Also please squash the commits and we are ready to merge

aledbf commented:
/approve

pcj commented:
OK, comments addressed & fixed, squashed into single commit and force pushed.  Should be ready for merge.  Thx.

antoineco commented on docs/examples/grpc/README.md:
````diff
@@ -0,0 +1,147 @@
+
+### Step 2: the kubernetes `Service`
+
+```yaml
````
Maybe a link to the corresponding YAML files would be better than copying their content directly here?
Or the other way around, keep the content of the manifests in the markdown but delete the files.

pcj replied:
Sure I can delete the redundant content.  I prefer the readability of the yaml in the markdown but could be convinced otherwise.  Unless another feels strongly, I'll just delete the yaml files.

ElvinEfendi replied:
I like the style in other docs where we provide complete `kubectl` command:
```
cat docs/examples/grpc/svc.yaml | kubectl apply --namespace=fortune-teller-app -f -
```

pcj replied:
@ElvinEfendi done.

antoineco commented on docs/examples/grpc/app.yaml:
```diff
@@ -0,0 +1,20 @@
+    spec:
+      containers:
+      - name: fortune-teller-app
+        image: container.stack.build/pcj/fortune-teller:latest
```
Should we host this in the repo together with echoserver and other tools @aledbf?

aledbf replied:
That makes sense

pcj replied:
I was planning on putting this in quay.io, but they still don't support docker registry v2.2 so I decided against it (ref https://github.com/bazelbuild/rules_docker/issues/102 gentle ping @philips).

```sh
E0502 07:46:27.210516   20444 docker_session_.py:335] Error during upload of: quay.io/pcj/fortune-teller:latest
F0502 07:46:27.210813   20444 fast_pusher_.py:162] Error publishing quay.io/pcj/fortune-teller:latest: response: {'status': '415', 'content-length': '131', 'server': 'nginx/1.13.6', 'connection': 'keep-alive', 'date': 'Wed, 02 May 2018 13:46:27 GMT', 'content-type': 'application/json'}
manifest invalid: {u'message': u'manifest schema version not supported'}
```

pcj replied:
@aledbf Is hosting this image as @antoineco suggests part of this PR?

aledbf replied:
Yes please

aledbf replied:
> Can I just mkdir images/grpc-fortune-teller

Yes. After merging the PR I will publish the image in the same place than the ingress image

pcj replied:
What will the anticipated image name be? (for the `app.yaml` image name).  Something under quay.io, gcr.io/google_containers, or elsewhere?

aledbf replied:
`quay.io/kubernetes-ingress-controller/grpc-fortune-teller:0.1`

pcj replied:
Done

pcj commented:
OK, I've added the grpc-fortune-teller under `images/`, updated image references to `quay.io/kubernetes-ingress-controller/grpc-fortune-teller:0.1`.

@aledbf, assuming you have bazel installed, you *should* be able to just type `$ bazel run //app:push` to push the image if your quay.io credentials are stored in `~/.docker/config.json` (docker login).  However, since quay.io does not support v2.2, it will likely fail with 'manifest version not supported'.  In this case you'll just have to `bazel run //app:image`, re-tag the image and push it manually via `docker push`.

pcj commented:
Unclear on why travis failed.

antoineco commented:
My general feeling is that we could definitely use a simpler example for demonstrating the use of gRPC. Fileutil.go and Bazel especially make me feel like we're showcasing a full-fledged Go project.

Regarding the content of the PR itself:
/lgtm

aledbf commented:
@pcj you need to add boilerplate code (license) to one or more files. Please execute `./hack/verify-boilerplate.sh` to see where's the issue

pcj commented:
Done.  Boilerplate added to *.go files and refactored fileutil away.

aledbf commented:
/lgtm
